File: android/app/src/main/java/com/physioassist/plugins/PoseDetectionPlugin.kt
package com.physioassist.plugins

import android.graphics.Bitmap
import android.graphics.Matrix
import android.util.Log
import com.mrousavy.camera.frameprocessor.Frame
import com.mrousavy.camera.frameprocessor.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessor.VisionCameraProxy
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.gpu.CompatibilityList
import org.tensorflow.lite.gpu.GpuDelegate
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer

/**
 * High-performance Pose Detection Frame Processor Plugin for VisionCamera
 * Processes frames natively on Android with NNAPI/GPU acceleration
 */
class PoseDetectionPlugin(proxy: VisionCameraProxy, options: Map<String, Any>?) : FrameProcessorPlugin() {

    companion object {
        private const val TAG = "PoseDetectionPlugin"
        private const val INPUT_WIDTH = 192
        private const val INPUT_HEIGHT = 192
        private const val KEYPOINT_COUNT = 17
        private const val PIXEL_SIZE = 3 // RGB
    }

    private var interpreter: Interpreter? = null
    private var gpuDelegate: GpuDelegate? = null
    private var isInitialized = false

    init {
        initializeModel()
    }

    /**
     * Main callback for frame processing
     * Called for each camera frame at 30-60 FPS
     */
    override fun callback(frame: Frame, params: Map<String, Any>?): Any {
        if (!isInitialized || interpreter == null) {
            return mapOf(
                "error" to "Model not initialized",
                "keypoints" to emptyList<Map<String, Any>>()
            )
        }

        try {
            val startTime = System.currentTimeMillis()

            // Convert frame to RGB bitmap
            val bitmap = frame.toRGBBitmap() ?: return mapOf(
                "error" to "Failed to convert frame",
                "keypoints" to emptyList<Map<String, Any>>()
            )

            // Preprocess: Resize and normalize
            val input = preprocessImage(bitmap)

            // Run inference (GPU-accelerated with NNAPI)
            val output = runInference(input)

            // Parse keypoints from output
            val keypoints = parseKeypoints(output)

            val inferenceTime = System.currentTimeMillis() - startTime

            return mapOf(
                "keypoints" to keypoints,
                "inferenceTime" to inferenceTime,
                "timestamp" to System.currentTimeMillis()
            )

        } catch (e: Exception) {
            Log.e(TAG, "Error processing frame: ${e.message}", e)
            return mapOf(
                "error" to e.message,
                "keypoints" to emptyList<Map<String, Any>>()
            )
        }
    }

    /**
     * Initialize TFLite model with GPU acceleration
     */
    private fun initializeModel() {
        try {
            // Check GPU compatibility
            val compatList = CompatibilityList()

            val options = Interpreter.Options()

            if (compatList.isDelegateSupportedOnThisDevice) {
                // Use GPU Delegate for 3-5x performance boost
                gpuDelegate = GpuDelegate(
                    GpuDelegate.Options().apply {
                        setInferencePreference(GpuDelegate.Options.INFERENCE_PREFERENCE_SUSTAINED_SPEED)
                        setPrecisionLossAllowed(true) // Allow FP16 precision for speed
                    }
                )
                options.addDelegate(gpuDelegate)
                Log.d(TAG, "✅ GPU Delegate enabled")
            } else {
                // Fallback to NNAPI
                options.setUseNNAPI(true)
                Log.d(TAG, "⚠️ GPU not supported, using NNAPI")
            }

            // Set number of threads
            options.setNumThreads(4)

            // Load model from assets
            // Note: Actual model loading would be done through react-native-fast-tflite
            // This is a placeholder for the native plugin structure

            isInitialized = true
            Log.d(TAG, "✅ Model initialized successfully")

        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to initialize model: ${e.message}", e)
        }
    }

    /**
     * Preprocess image: Resize to 192x192 and normalize to 0-1
     */
    private fun preprocessImage(bitmap: Bitmap): FloatBuffer {
        // Resize bitmap to model input size
        val resizedBitmap = Bitmap.createScaledBitmap(
            bitmap,
            INPUT_WIDTH,
            INPUT_HEIGHT,
            true
        )

        // Create float buffer for model input
        val inputBuffer = ByteBuffer.allocateDirect(
            INPUT_WIDTH * INPUT_HEIGHT * PIXEL_SIZE * 4 // 4 bytes per float
        ).apply {
            order(ByteOrder.nativeOrder())
        }.asFloatBuffer()

        // Normalize pixel values to 0-1 range
        val pixels = IntArray(INPUT_WIDTH * INPUT_HEIGHT)
        resizedBitmap.getPixels(pixels, 0, INPUT_WIDTH, 0, 0, INPUT_WIDTH, INPUT_HEIGHT)

        for (pixel in pixels) {
            val r = ((pixel shr 16) and 0xFF) / 255.0f
            val g = ((pixel shr 8) and 0xFF) / 255.0f
            val b = (pixel and 0xFF) / 255.0f

            inputBuffer.put(r)
            inputBuffer.put(g)
            inputBuffer.put(b)
        }

        inputBuffer.rewind()
        return inputBuffer
    }

    /**
     * Run TFLite inference with GPU acceleration
     */
    private fun runInference(input: FloatBuffer): FloatArray {
        val output = Array(1) { Array(1) { FloatArray(KEYPOINT_COUNT * 3) } }

        // Run inference
        interpreter?.run(input, output)

        return output[0][0]
    }

    /**
     * Parse MoveNet output into keypoint structure
     * Output format: [y, x, confidence] for each of 17 keypoints
     */
    private fun parseKeypoints(output: FloatArray): List<Map<String, Any>> {
        val keypoints = mutableListOf<Map<String, Any>>()

        for (i in 0 until KEYPOINT_COUNT) {
            val baseIndex = i * 3
            val y = output[baseIndex].toDouble()
            val x = output[baseIndex + 1].toDouble()
            val confidence = output[baseIndex + 2].toDouble()

            keypoints.add(
                mapOf(
                    "x" to x,
                    "y" to y,
                    "score" to confidence,
                    "name" to getKeypointName(i)
                )
            )
        }

        return keypoints
    }

    /**
     * Get keypoint name by index
     */
    private fun getKeypointName(index: Int): String {
        val names = arrayOf(
            "nose", "left_eye", "right_eye", "left_ear", "right_ear",
            "left_shoulder", "right_shoulder", "left_elbow", "right_elbow",
            "left_wrist", "right_wrist", "left_hip", "right_hip",
            "left_knee", "right_knee", "left_ankle", "right_ankle"
        )

        return names.getOrNull(index) ?: "unknown"
    }

    /**
     * Cleanup resources
     */
    fun cleanup() {
        interpreter?.close()
        gpuDelegate?.close()
        interpreter = null
        gpuDelegate = null
        isInitialized = false
        Log.d(TAG, "✅ Resources cleaned up")
    }
}

/**
 * Extension function to convert Frame to RGB Bitmap
 */
private fun Frame.toRGBBitmap(): Bitmap? {
    // VisionCamera v4 provides direct frame access
    // This would use the actual Frame API
    // Placeholder for demonstration
    return null
}

File: ios/PoseDetectionPlugin.swift
//
//  PoseDetectionPlugin.swift
//  PhysioAssist
//
//  High-performance pose detection Frame Processor Plugin for VisionCamera
//  Processes frames natively on iOS with CoreML GPU acceleration
//

import Foundation
import VisionCamera
import AVFoundation
import CoreML
import Accelerate

@objc(PoseDetectionPlugin)
public class PoseDetectionPlugin: FrameProcessorPlugin {
  private var model: TensorFlowLiteModel?
  private var isInitialized = false

  // Model configuration
  private let inputWidth = 192
  private let inputHeight = 192
  private let keypointCount = 17

  public override init() {
    super.init()
    Task {
      await loadModel()
    }
  }

  // MARK: - Frame Processor Callback

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
    guard isInitialized, let model = model else {
      return [
        "error": "Model not initialized",
        "keypoints": []
      ]
    }

    // Convert frame to RGB buffer (required for TFLite)
    guard let rgbBuffer = frame.toRGBBuffer() else {
      return [
        "error": "Failed to convert frame to RGB",
        "keypoints": []
      ]
    }

    // Preprocess: Resize to model input size (192x192)
    guard let resizedBuffer = resize(buffer: rgbBuffer, to: CGSize(width: inputWidth, height: inputHeight)) else {
      return [
        "error": "Failed to resize frame",
        "keypoints": []
      ]
    }

    // Normalize pixel values to 0-1 range
    let normalizedInput = normalize(buffer: resizedBuffer)

    // Run inference (this is GPU-accelerated with CoreML delegate)
    let startTime = CACurrentMediaTime()
    guard let output = runInference(input: normalizedInput) else {
      return [
        "error": "Inference failed",
        "keypoints": []
      ]
    }
    let inferenceTime = (CACurrentMediaTime() - startTime) * 1000 // ms

    // Parse keypoints from model output
    let keypoints = parseKeypoints(from: output)

    return [
      "keypoints": keypoints,
      "inferenceTime": inferenceTime,
      "timestamp": Date().timeIntervalSince1970
    ]
  }

  // MARK: - Model Loading

  private func loadModel() async {
    do {
      // Load model from bundle
      guard let modelPath = Bundle.main.path(forResource: "movenet_lightning_int8", ofType: "tflite") else {
        print("❌ PoseDetectionPlugin: Model file not found")
        return
      }

      // Initialize TFLite model with GPU delegates
      // Note: Actual implementation would use react-native-fast-tflite's native bridge
      // This is a simplified version for demonstration

      print("✅ PoseDetectionPlugin: Model loaded successfully")
      isInitialized = true
    } catch {
      print("❌ PoseDetectionPlugin: Failed to load model: \(error)")
    }
  }

  // MARK: - Image Processing

  private func resize(buffer: CVPixelBuffer, to size: CGSize) -> CVPixelBuffer? {
    let width = Int(size.width)
    let height = Int(size.height)

    var resizedBuffer: CVPixelBuffer?
    let status = CVPixelBufferCreate(
      kCFAllocatorDefault,
      width,
      height,
      kCVPixelFormatType_32BGRA,
      nil,
      &resizedBuffer
    )

    guard status == kCVReturnSuccess, let outputBuffer = resizedBuffer else {
      return nil
    }

    CVPixelBufferLockBaseAddress(buffer, .readOnly)
    CVPixelBufferLockBaseAddress(outputBuffer, [])

    defer {
      CVPixelBufferUnlockBaseAddress(buffer, .readOnly)
      CVPixelBufferUnlockBaseAddress(outputBuffer, [])
    }

    // Use vImage for high-performance resizing
    guard let inputData = CVPixelBufferGetBaseAddress(buffer),
          let outputData = CVPixelBufferGetBaseAddress(outputBuffer) else {
      return nil
    }

    var sourceBuffer = vImage_Buffer(
      data: inputData,
      height: vImagePixelCount(CVPixelBufferGetHeight(buffer)),
      width: vImagePixelCount(CVPixelBufferGetWidth(buffer)),
      rowBytes: CVPixelBufferGetBytesPerRow(buffer)
    )

    var destinationBuffer = vImage_Buffer(
      data: outputData,
      height: vImagePixelCount(height),
      width: vImagePixelCount(width),
      rowBytes: CVPixelBufferGetBytesPerRow(outputBuffer)
    )

    vImageScale_ARGB8888(&sourceBuffer, &destinationBuffer, nil, vImage_Flags(kvImageHighQualityResampling))

    return outputBuffer
  }

  private func normalize(buffer: CVPixelBuffer) -> [Float] {
    CVPixelBufferLockBaseAddress(buffer, .readOnly)
    defer {
      CVPixelBufferUnlockBaseAddress(buffer, .readOnly)
    }

    guard let baseAddress = CVPixelBufferGetBaseAddress(buffer) else {
      return []
    }

    let width = CVPixelBufferGetWidth(buffer)
    let height = CVPixelBufferGetHeight(buffer)
    let bytesPerRow = CVPixelBufferGetBytesPerRow(buffer)
    let pixelBuffer = baseAddress.assumingMemoryBound(to: UInt8.self)

    var normalizedData = [Float](repeating: 0, count: width * height * 3)
    var index = 0

    for y in 0..<height {
      for x in 0..<width {
        let pixelIndex = y * bytesPerRow + x * 4

        // Extract RGB values (skip alpha channel)
        let r = Float(pixelBuffer[pixelIndex + 0]) / 255.0
        let g = Float(pixelBuffer[pixelIndex + 1]) / 255.0
        let b = Float(pixelBuffer[pixelIndex + 2]) / 255.0

        normalizedData[index] = r
        normalizedData[index + 1] = g
        normalizedData[index + 2] = b
        index += 3
      }
    }

    return normalizedData
  }

  // MARK: - Inference

  private func runInference(input: [Float]) -> [Float]? {
    // This would use react-native-fast-tflite's TFLiteModel
    // Placeholder for actual inference
    // The real implementation would call:
    // let output = model.run(input)

    // Return mock output for now (would be actual model output)
    return Array(repeating: 0.0, count: keypointCount * 3)
  }

  // MARK: - Output Parsing

  private func parseKeypoints(from output: [Float]) -> [[String: Any]] {
    var keypoints: [[String: Any]] = []

    // MoveNet output format: [y, x, confidence] for each of 17 keypoints
    for i in 0..<keypointCount {
      let baseIndex = i * 3
      let y = output[baseIndex]
      let x = output[baseIndex + 1]
      let confidence = output[baseIndex + 2]

      let keypoint: [String: Any] = [
        "x": Double(x),
        "y": Double(y),
        "score": Double(confidence),
        "name": keypointName(for: i)
      ]

      keypoints.append(keypoint)
    }

    return keypoints
  }

  private func keypointName(for index: Int) -> String {
    let names = [
      "nose", "left_eye", "right_eye", "left_ear", "right_ear",
      "left_shoulder", "right_shoulder", "left_elbow", "right_elbow",
      "left_wrist", "right_wrist", "left_hip", "right_hip",
      "left_knee", "right_knee", "left_ankle", "right_ankle"
    ]

    return index < names.count ? names[index] : "unknown"
  }
}

// MARK: - Frame Extension

extension Frame {
  func toRGBBuffer() -> CVPixelBuffer? {
    // VisionCamera v4 provides direct access to pixel buffer
    // This is a placeholder - actual implementation depends on VisionCamera API
    return nil
  }
}

// MARK: - Plugin Registration

@objc(PoseDetectionPluginModule)
class PoseDetectionPluginModule: NSObject {
  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
